File: packages/engine/src/registry/nouns.ts
/**
 * The ontology register: every THING this world is made of, declared with where it lives.
 *
 * @spec Law 2 Law 4 Law 15 XI-14 Appendix A Appendix C
 *
 * `ParamRegister` forces every declared NUMBER to name its kind, its unit and its owner, and the
 * build fails if one does not. There was no equivalent for CATEGORIES. The specification's systems
 * enumerate BEHAVIOURS and `registry/kinds.ts` enumerates SETTLEMENT OBJECTS; nothing enumerated the
 * things the economy is made of. So "is there a noun for this?" had no lookup — it had a search, and
 * a search finds only what somebody already tried to build. Sixteen missing primitives were found
 * over six passes of reading, each pass finding more only because the method changed.
 *
 * `MechanismContext.state()` is where a module puts something the kernel has no home for. It is a
 * bag: it takes a name and an object and asks nothing. Seventeen modules keep nineteen slots in it,
 * and its own docstring lists the missing primitives without noticing — "a register of employment
 * rows, a book of invoices, a party's outlooks". Each of those is an economic noun with no kernel
 * store, kept privately, invisible to the audit, to the estate's ranking and to every other module.
 *
 * SO A SLOT IS DECLARED OR IT DOES NOT OPEN. A module says what is in it and which of three things
 * it is, and the one that means "this is a noun the kernel should own" must name the plan item that
 * gives it that home — the same construction `ParamRegister` uses for a placeholder, for the same
 * reason: a stand-in with no scheduled death is a permanent one (Law 2).
 *
 * The count of nouns still in a bag is the honest measure of how much ontology is missing, and it is
 * reported rather than hidden (Appendix C).
 */
import { InvalidRegistry } from '../core/errors.js';

/**
 * Law 2: what a module's own store IS, from a closed list.
 *
 * `noun` is a thing this economy HAS, kept in a module's bag because the kernel has no store for it
 * — an employment, a lease, an invoice, a party's outlook. It is a placeholder for a kernel noun and
 * names the plan item that gives it one.
 *
 * `working` is state a phase hands to a later phase WITHIN a period and nothing outside the module
 * has any business reading: a draw the kernel allowed that is waiting to become a row, a line whose
 * wipe has been announced so it is not announced twice. It is not a fact about the world, it is how
 * one module gets from one of its phases to the next.
 *
 * `physics` is the module's own subject matter, which is private by right and wants no kernel home:
 * the weather is the environment module's, and nothing else in this world has an opinion about it.
 */
export type NounKind = 'noun' | 'working' | 'physics';

export const NOUN_KINDS: readonly NounKind[] = ['noun', 'working', 'physics'];

/** Where a noun kept in a module's bag is going, and what ends the arrangement (Law 2). */
export interface NounHome {
  /** The kernel noun it is an instance of, named as `docs/IMPLEMENTATION.md` names it. */
  readonly noun: string;
  /** The plan item that builds that noun and deletes this declaration in the same change. */
  readonly planItem: string;
}

/** What a module declares about one of its own stores. The owner is stamped at assembly (Law 4). */
export interface NounEntry {
  /** The name the module passes to `ctx.state`. */
  readonly name: string;
  readonly kind: NounKind;
  /** What is in it, in the terms a reader of the economy would use. */
  readonly holds: string;
  /** Why it is this kind rather than another (Law 16: a comment says what the artefact cannot). */
  readonly why: string;
  readonly standsInFor?: NounHome;
}

export interface NounDecl extends NounEntry {
  /** The module that keeps it. Stamped by `assemble`, never written by the module itself. */
  readonly owner: string;
}

export interface OntologyReport {
  readonly counts: Readonly<Record<NounKind, number>>;
  /** Every noun still living in a module's bag, and the item that gives it a kernel home. */
  readonly homeless: readonly { owner: string; name: string; noun: string; planItem: string }[];
}

const key = (owner: string, name: string): string => `${owner}/${name}`;

export class OntologyRegister {
  private readonly decls: ReadonlyMap<string, NounDecl>;

  constructor(decls: readonly NounDecl[]) {
    const map = new Map<string, NounDecl>();
    for (const d of decls) {
      const k = key(d.owner, d.name);
      if (map.has(k)) throw new InvalidRegistry('Law 4', `store ${k} is declared twice`);
      if (d.name.length === 0)
        throw new InvalidRegistry('Law 15', `${d.owner} declares a store with no name`);
      if (d.holds.length === 0)
        throw new InvalidRegistry('Law 16', `store ${k} does not say what is in it`);
      if (d.why.length === 0) throw new InvalidRegistry('Law 16', `store ${k} has no reason`);
      if (d.kind === 'noun' && d.standsInFor === undefined) {
        throw new InvalidRegistry(
          'Law 2',
          `${k} is declared a noun and names no plan item that gives it a kernel home; a stand-in with no scheduled death is a permanent one`,
        );
      }
      if (d.kind !== 'noun' && d.standsInFor !== undefined) {
        throw new InvalidRegistry(
          'Law 2',
          `${k} names a kernel home but is declared ${d.kind}; only a noun is going somewhere`,
        );
      }
      map.set(k, d);
    }
    this.decls = map;
  }

  /**
   * The declaration for a store, or a refusal.
   *
   * This is the whole point of the register: a module reaching for a category the kernel has no
   * home for gets a build failure that says so, instead of reaching for the nearest bag — which is
   * the move that produced every missing primitive in `docs/IMPLEMENTATION.md`.
   */
  declared(owner: string, name: string): NounDecl {
    const d = this.decls.get(key(owner, name));
    if (d === undefined) {
      throw new InvalidRegistry(
        'Law 15',
        `${owner} keeps a store called "${name}" that it never declared. Say what is in it and which of ${NOUN_KINDS.join(' | ')} it is; a noun names the plan item that gives it a kernel home (docs/RECORD.md item 0)`,
      );
    }
    return d;
  }

  all(): readonly NounDecl[] {
    return [...this.decls.values()];
  }

  /** Appendix C: recount rather than adjust. What is still in a bag, and what takes it out. */
  report(): OntologyReport {
    const counts: Record<NounKind, number> = { noun: 0, working: 0, physics: 0 };
    const homeless: { owner: string; name: string; noun: string; planItem: string }[] = [];
    for (const d of this.decls.values()) {
      counts[d.kind] += 1;
      const home = d.standsInFor;
      if (home !== undefined) {
        homeless.push({ owner: d.owner, name: d.name, noun: home.noun, planItem: home.planItem });
      }
    }
    return { counts, homeless };
  }
}

File: packages/kernel-rs/src/nouns.rs
//! The ontology register: every store a module keeps, declared for what it is.
//!
//! It is to categories what `params` is to numbers. A module keeps state between its phases, and
//! the question nobody could answer was which of those stores are FACTS ABOUT THE WORLD that the
//! kernel should own, and which are a module's own scratch. An undeclared store is refused at the
//! read, so the answer cannot be avoided by not writing it down.
//!
//! Its count of HOMELESS nouns is the honest measure of how much ontology is missing — a noun
//! is a fact about the world, so it belongs in a kernel store, and one that has not got there yet
//! names the plan item that will give it a home. The count must fall; it is not a number to
//! tolerate.

use std::collections::HashMap;

#[derive(Clone, PartialEq, Eq, Debug)]
pub enum Sort {
    /// A FACT ABOUT THE WORLD, which belongs in a kernel store. Until it is there it names the
    /// plan item that gives it a home, and it is HOMELESS — which is what the count counts.
    Noun { home: Option<String> },
    /// A counter or a plan within one module's own phase, which does not survive in any sense a
    /// reader could use. It is not a fact about the world.
    Working,
    /// Weather, geography, the grid: given, not decided by anybody in this world.
    Physics,
}

pub struct NounDecl {
    pub name: String,
    pub sort: Sort,
    /// What it holds, in a reader's words.
    pub holds: String,
    /// Why it is the sort it is. A noun with no reason is a store nobody classified.
    pub why: String,
}

#[derive(Default)]
pub struct Nouns {
    at: HashMap<String, usize>,
    sort: Vec<Sort>,
    holds: Vec<String>,
    why: Vec<String>,
}

impl Nouns {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn declare(&mut self, d: NounDecl) {
        assert!(!self.at.contains_key(&d.name), "Law 4: the store {} is declared twice", d.name);
        assert!(!d.why.is_empty(), "Law 16: {} is declared with no reason", d.name);
        if let Sort::Noun { home: Some(item) } = &d.sort {
            assert!(!item.is_empty(), "a noun's home names the item that gives it one");
        }
        self.at.insert(d.name.clone(), self.sort.len());
        self.sort.push(d.sort);
        self.holds.push(d.holds);
        self.why.push(d.why);
    }

    /// The read a module's store goes through. An undeclared store is refused HERE, at the read,
    /// because that is the moment the omission exists.
    pub fn sort_of(&self, name: &str) -> &Sort {
        match self.at.get(name) {
            Some(&at) => &self.sort[at],
            None => panic!(
                "Law 15: the store {name} is not declared — every store is a noun, a working store or physics"
            ),
        }
    }

    /// The nouns with no kernel home yet, in order. This count is the measure, and it must fall.
    pub fn homeless(&self) -> Vec<(&str, &str)> {
        let mut out: Vec<(&str, &str)> = Vec::new();
        for (name, &at) in &self.at {
            if let Sort::Noun { home: Some(item) } = &self.sort[at] {
                out.push((name.as_str(), item.as_str()));
            }
        }
        out.sort_by(|a, b| a.0.cmp(b.0));
        out
    }

    pub fn len(&self) -> usize {
        self.sort.len()
    }

    pub fn is_empty(&self) -> bool {
        self.sort.is_empty()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn decl(name: &str, sort: Sort) -> NounDecl {
        NounDecl {
            name: name.to_string(),
            sort,
            holds: "something".to_string(),
            why: "because the test says so".to_string(),
        }
    }

    #[test]
    fn the_homeless_count_is_the_measure_and_a_settled_noun_is_not_in_it() {
        let mut n = Nouns::new();
        n.declare(decl("control.advisory", Sort::Working));
        n.declare(decl("weather.today", Sort::Physics));
        n.declare(decl("firm.order.book", Sort::Noun { home: Some("13k".to_string()) }));
        n.declare(decl("register.holdings", Sort::Noun { home: None }));
        assert_eq!(n.len(), 4);
        // Only the noun that has not reached a kernel store is homeless.
        assert_eq!(n.homeless(), vec![("firm.order.book", "13k")]);
    }

    #[test]
    #[should_panic(expected = "is not declared")]
    fn an_undeclared_store_is_refused_at_the_read() {
        let n = Nouns::new();
        n.sort_of("somebody.kept.this.quietly");
    }

    #[test]
    #[should_panic(expected = "is declared twice")]
    fn one_store_has_one_declaration() {
        let mut n = Nouns::new();
        n.declare(decl("x", Sort::Working));
        n.declare(decl("x", Sort::Working));
    }
}
